### PR TITLE
Task#16 - Feat:  Show product details component 

### LINKED ISSUE #63: UX: Filter product card
- [x] #64
- [x] #66
- [x] #72
- [x] #77

### LINKED ISSUE #77: Task#16 - DevTask: Update product details show
## OVERVIEW
- When use click products card show details for products
- Can show product details in Products list and Favorate list
- Can close popover product details and chosse another product details

### PR DESCRIPTION
**Pull request by TheHienDuong**
- Closes #77 

## OVERVIEW
- Update popover for product details
- Add use state to select product
- handle selecting and clossing products card
- Update select pro prop for components
- Update comment and propTypes

## DEMO

https://github.com/Agility-Internship/Duong-The-Hien-ReactJS/assets/113186169/4f7117f9-d85b-4d3e-8a8c-faac0c3d39c8

### CODE REVIEW

File: practice-two/src/App.jsx [hunk 1]
```diff
@@ -25,6 +25,24 @@ const App = () => {
...
   const [favoriteProductIDs, setFavoriteProductIDs] = useState([]);
 
+  // State for storing the selected product's ID for details view
+  const [selectedProductID, setSelectedProductID] = useState(null);
+
+  // Function to handle selecting a product card
+  const handleSelectProduct = (productID) => {
+    setSelectedProductID((prevProductID) => {
+      if (prevProductID === productID) {
+        return null; // Close the details if it's the same product
```
Comment (phattruong-agilityio) [inline practice-two/src/App.jsx]: `return null;` then the comment is "Close the details if it's the same product"? Does it make sense ? There are two different actions

File: practice-two/src/App.jsx [hunk 2]
```diff
@@ -25,6 +25,24 @@ const App = () => {
...
+      }
+      return productID; // Display the details of the selected product
```
Comment (phattruong-agilityio) [inline practice-two/src/App.jsx]: the same here. I don't see it's relevant to "Display the details of the selected product"

File: practice-two/src/App.jsx [hunk 3]
```diff
@@ -25,6 +25,24 @@ const App = () => {
   // State for storing the favorite product IDs
   const [favoriteProductIDs, setFavoriteProductIDs] = useState([]);
 
+  // State for storing the selected product's ID for details view
+  const [selectedProductID, setSelectedProductID] = useState(null);
+
+  // Function to handle selecting a product card
```
Comment (phattruong-agilityio) [inline practice-two/src/App.jsx]: Check the comment format
it's should be 
```
/**
 * function do sth
 * @params ....
 * @return ....
 */
```

right? Did I told you before

File: practice-two/src/App.jsx [hunk 4]
```diff
@@ -25,6 +25,24 @@ const App = () => {
...
+  const handleSelectProduct = (productID) => {
+    setSelectedProductID((prevProductID) => {
+      if (prevProductID === productID) {
+        return null; // Close the details if it's the same product
+      }
+      return productID; // Display the details of the selected product
+    });
+  };
+
+  // Callback for closing the product details
```
Comment (phattruong-agilityio) [inline practice-two/src/App.jsx]: check the comment format and the content. I don't see any "callback"? do you understand the "callback"?

File: practice-two/src/App.jsx [hunk 5]
```diff
@@ -25,6 +25,24 @@ const App = () => {
...
+  const handleCloseProductDetails = () => {
+    setSelectedProductID(null);
+  };
+
   // Function to handle toggling a product as favorite
```
Comment (phattruong-agilityio) [inline practice-two/src/App.jsx]: The same here

File: practice-two/src/App.jsx [hunk 6]
```diff
@@ -107,16 +125,27 @@ const App = () => {
...
           favoriteProductIDs={favoriteProductIDs}
           onToggleProductFavorite={toggleProductFavorite}
+          onSelectProduct={handleSelectProduct}
         />
       </section>
-      <ProductDetails product={allProducts[56]} />
+      <section>
+        {selectedProductID !== null && (
+          <ProductDetails
+            product={allProducts.find((product) => product.id === selectedProductID)}
```
Comment (phattruong-agilityio) [inline practice-two/src/App.jsx]: Please don't do compute in render jsx.

```jsx
const selectedProduct = allProducts.find((product) => product.id === selectedProductID);

...

return (
   ...
   <{selectedProductID !== null && (
          <ProductDetails
            product={selectedProduct}
            favoriteProductIDs={favoriteProductIDs}
            onClose={handleCloseProductDetails}
            onToggleProductFavorite={toggleProductFavorite}
  />
)
```

File: practice-two/src/components/ProductCard.jsx [hunk 7]
```diff
@@ -13,14 +13,26 @@ import FavoriteButton from './common/Button/ButtonFavorite';
...
  * @returns {JSX.Element} Product Card content.
  */
-function ProductCard({ product = [], isFavoriteProduct = [], onSelectFavorite = () => {} }) {
+function ProductCard({
+  product = [],
+  favoriteProductIDs = [],
+  onSelectFavorite = () => {},
+  onSelectProduct = () => {},
+}) {
+  // Function to handle get ID when a favorite product is selected
```
Comment (phattruong-agilityio) [inline practice-two/src/components/ProductCard.jsx]: check comment format here

File: practice-two/src/components/ProductCard.jsx [hunk 8]
```diff
@@ -13,14 +13,26 @@ import FavoriteButton from './common/Button/ButtonFavorite';
...
   const handleFavoriteToggle = () => {
     onSelectFavorite(product.id);
   };
+
+  // Function to handle get ID when a product card is selected
+  const handleSelectProduct = () => {
```
Comment (phattruong-agilityio) [inline practice-two/src/components/ProductCard.jsx:39]: check comment format here

File: practice-two/src/components/ProductDetails.jsx [hunk 9]
```diff
@@ -1,80 +1,120 @@
...
+    <section
+      className={`fixed inset-0 bg-opacity-70 backdrop-filter backdrop-blur-lg flex items-center justify-center z-50 ${
+        isPopoverOpen ? 'bg-gray-800' : 'hidden'
+      }`}
+    >
+      <div className="container min-h-screen w-[40%] h-[30%] flex justify-center items-center">
+        <Popover
+          closeButtonoo
+          isOpen={isPopoverOpen}
+          onClose={handleButtonClick}
```
Comment (phattruong-agilityio) [inline practice-two/src/components/ProductDetails.jsx]: check the naming here

File: practice-two/src/components/ProductDetails.jsx [hunk 10]
```diff
@@ -1,80 +1,120 @@
...
+          popoverClassName="relative bg-white rounded-2xl max-h-[700x] overflow-x-auto shadow-2xl p-10 m-2 "
+          arrowPopover={false}
+        >
+          <div className="grid grid-cols-1 lg:grid-cols-2 grid-rows-none items-center">
+            <div className="grid grid-cols-3 grid-rows-auto gap-3 mr-5">
+              {imageUrls.map((imageUrl, index) => (
+                <div
+                  // eslint-disable-next-line react/no-array-index-key
+                  key={index}
```
Comment (phattruong-agilityio) [inline practice-two/src/components/ProductDetails.jsx:64]: key={index} huh?

File: practice-two/src/components/ProductDetails.jsx [hunk 11]
```diff
@@ -1,80 +1,120 @@
...
+                  className={`border rounded-2xl ${index === 0 ? 'active col-span-3' : ''}`}
+                >
+                  <CardImage src={imageUrl} alt={`Product Image ${index + 1}`} />
```
Comment (phattruong-agilityio) [inline practice-two/src/components/ProductDetails.jsx]: "Product Image 1", "Product Image 2", "Product Image 3" does this alt make sense??

File: practice-two/src/components/ProductDetails.jsx [hunk 12]
```diff
@@ -1,80 +1,120 @@
...
+              ))}
             </div>
-            <div className="inline-block">
-              {selectedProduct.installment && (
-                <Typography color="black" size="md" variant="solid">
-                  {selectedProduct.installment}
+            <div>
+              <div className="basic-info flex flex-col gap-4 relative">
+                <div className="flex justify-between hover:text-blue-700">
+                  <Link href="/product-card" underline={false} size="xl" variant="custom-variant">
```
Comment (phattruong-agilityio) [inline practice-two/src/components/ProductDetails.jsx]: why have href here

File: practice-two/src/components/layout/ListProductsCard.jsx [hunk 13]
```diff
@@ -13,13 +13,24 @@ import Typography from '../common/Typography';
...
+  favoriteProductIDs = [],
+  onToggleProductFavorite = () => {},
+  onSelectProduct = () => {},
+}) => {
   // Function to handle the click event when a favorite product is selected
   const handleFavoriteSelect = (productId) => {
     onToggleProductFavorite(productId);
   };
+
+  // Function to handle the click event when a product card is selected
```
Comment (phattruong-agilityio) [inline practice-two/src/components/layout/ListProductsCard.jsx]: The same here

Comment (phattruong-agilityio) [reply practice-two/src/components/layout/ListProductsCard.jsx]: Update ALL similer places

File: practice-two/src/App.jsx (hunk 1 above)
Comment (TheHienDuong) [reply practice-two/src/App.jsx]: Yes, Done

File: practice-two/src/App.jsx (hunk 2 above)
Comment (TheHienDuong) [reply practice-two/src/App.jsx]: Done

File: practice-two/src/App.jsx (hunk 3 above)
Comment (TheHienDuong) [reply practice-two/src/App.jsx]: yes you said, updated the functions

File: practice-two/src/App.jsx (hunk 4 above)
Comment (TheHienDuong) [reply practice-two/src/App.jsx]: Done

File: practice-two/src/App.jsx (hunk 5 above)
Comment (TheHienDuong) [reply practice-two/src/App.jsx]: Done

File: practice-two/src/App.jsx (hunk 6 above)
Comment (TheHienDuong) [reply practice-two/src/App.jsx]: Done

File: practice-two/src/components/ProductCard.jsx (hunk 7 above)
Comment (TheHienDuong) [reply practice-two/src/components/ProductCard.jsx]: Done

File: practice-two/src/components/ProductCard.jsx (hunk 8 above)
Comment (TheHienDuong) [reply practice-two/src/components/ProductCard.jsx:39]: Done

File: practice-two/src/components/ProductDetails.jsx (hunk 12 above)
Comment (TheHienDuong) [reply practice-two/src/components/ProductDetails.jsx]: This is the link to the product that is the same as the one in the Products Card

File: practice-two/src/components/layout/ListProductsCard.jsx (hunk 13 above)
Comment (TheHienDuong) [reply practice-two/src/components/layout/ListProductsCard.jsx]: Done

File: practice-two/src/components/ProductDetails.jsx (hunk 10 above)
Comment (TheHienDuong) [reply practice-two/src/components/ProductDetails.jsx:64]: Yes, I think so, corresponding to the values ​​of property images

File: practice-two/src/components/ProductDetails.jsx (hunk 11 above)
Comment (TheHienDuong) [reply practice-two/src/components/ProductDetails.jsx]: Yes, I udpated into Color '..' styles for phones

File: practice-two/src/components/ProductDetails.jsx (hunk 9 above)
Comment (TheHienDuong) [reply practice-two/src/components/ProductDetails.jsx]: onClosePopover udpated

**Pull request merged**